### DISCUSSION TITLE
Optimizing Vineyard for big payloads

### DISCUSSION THREAD

**Discussion by dv-rastogi** [v6d-io/v6d#1419] [General]
Hi, 
We are using Vineyard to write & read big payloads (~GB). Are there any configurations / tweaks one can use for maximizing performance there? For example, maybe one can optimize the underlying memory allocator for allocating much larger blocks thus reducing the allocator overhead?

**Comment by sighingnow** [comment]
Hi @watch24hrs-iiitd,

Do you mean single large blobs or many small blobs but the summarization of size can be large?

**Reply by dv-rastogi** [reply]
Hi,
I meant single large blobs.

**Reply by sighingnow** [reply]
Sorry for missing this message.

**Comment by dv-rastogi** [comment]
Hi, 

Any suggestions for the above?

**Reply by sighingnow** [reply]
For single large blobs, the main cost of creating blobs is allocating physical pages (the first run) and memcpy.

- for the first issue, vineyard supports a command line flags `-reserve_memory` to pre-allocate physical pages from OS, for a fresh launched vineyardd, this flag can reduce the time of creating 8G blob from `3.6s` to `2.8s`.

- for the second issue, we haven't invested much to provide memcpy utilities yet, but that could be optimized by avoiding the memcpy by refactoring to first creating blob builders and filling you logic data into the blob to avoid the extra memcpy.

**Reply by sighingnow** [reply]
We have provided a client-side allocator as well to allow applications allocate memory directly on vineyard's shared memory region but that are still in very early stage, see also 

- https://github.com/v6d-io/v6d/blob/main/modules/malloc/malloc_wrapper.h
- https://github.com/v6d-io/v6d/blob/main/modules/malloc/mimalloc_allocator.h

**Reply by dv-rastogi** [reply]
Thank you, will look into this!

**Comment by dv-rastogi** [comment]
Furthermore, can we optimize anything on the read end for large payloads?

**Reply by sighingnow** [reply]
Read blob from vineyard is **zero-copy** with `mmap()` so getting large blob from vineyard should yields no extra overhead.

**Reply by sighingnow** [reply]
If the get blob performance is a bottleneck in your case, you can share the case and we could take a look into the trouble.

**Reply by dv-rastogi** [reply]
Read isn't a bottleneck in our case, although we were just looking if anything specific could be done to boost it even further. Thank you for your help!
